lib/Data/Aeson/Combinators/Decode.hs

```diff
@@ -607,6 +611,51 @@ path :: JSONPath -> Decoder a -> Decoder a
 path pth d = foldr element d pth
 {-# INLINE path #-}
 
+-- | Try a decoder and get back a 'Just a' if it succeeds and 'Nothing' if it fails.
+-- In other words, this decoder always succeeds with a 'Maybe a' value.
+--
+-- > >>> decode (jsonMaybe string) "42"
+-- > Just Nothing
+-- > >>> decode (jsonMaybe int) "42"
+-- > Just (Just 42)
+jsonMaybe :: Decoder a -> Decoder (Maybe a)
+jsonMaybe (Decoder d) =
+  Decoder $ \val ->
+    case parse d val of
+      Success x -> return (Just x)
+      Error _ -> return Nothing
+{-# INLINE jsonMaybe #-}
+
+-- | Try a decoder and get back a 'Right a' if it succeeds and a 'Left String' if it fails.
+-- In other words, this decoder always succeeds with an 'Either String a' value.
+--
+-- > >>> decode (jsonEither string) "42"
+-- > Just (Left "expected String, but encountered Number")
+-- > >>> decode (jsonEither int) "42"
+-- > Just (Right 42)
+jsonEither :: Decoder a -> Decoder (Either String a)
+jsonEither (Decoder d) =
+  Decoder $ \val ->
+    case parse d val of
+      Success x -> return (Right x)
+      Error err -> return (Left err)
+{-# INLINE jsonEither #-}
+
+-- | Try a number of decoders in order and return the first success.
+--
+-- > >>> decode (oneOf [words <$> string, list string]) "\"Hello world!\""
+-- > Just ["Hello", "world!"]
+-- > >>> decode (oneOf [words <$> string, list string]) "[\"Hello world!\"]"
+-- > Just ["Hello world!"]
+-- > >>> decode (oneOf [Right <$> bool, return (Left "Not a boolean")]) "false"
+-- > Just (Right False)
+-- > >>> decode (oneOf [Right <$> bool, return (Left "Not a boolean")]) "42"
+-- > Just (Left "Not a boolean")
+oneOf :: [Decoder a] -> Decoder a
+oneOf [] = Fail.fail "No decoders provided."
+oneOf decoders =
+  foldr1 (<|>) decoders
+{-# INLINE oneOf #-}
```

Comment:
This seems to me re-implementation of [asum function from base](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Foldable.html#v:asum) can you elaborate why you don't want to use that (I feel there might be some good reason I'm missing).

Reply:
No reason! I did not know about `asum`, actually. I'm still new to Haskell but I'm very familiar with Elm Decoders, so I was just looking for what I already knew.

Reply:
No worries. Let's keep that function for a now. There will be at least difference in error handling because `asum` will fail with less custom error (https://github.com/turboMaCk/aeson-combinators/pull/10/files#diff-58641edeed79657d5a443c79e5c55b17R280) so I think this is worth at least considering nonetheless.

Reply:
I think we can include this function. Though I would rather avoid the error case of being possible (as an improvement to elm) and use `Data.List.NonEmpty (Decoder a)` instead of `[ Decoder a ]`.